### ISSUE TITLE
Release Manager cannot find param in .env files in ODS 3.x

### ISSUE THREAD

**Issue by felipecruz91** [opendevstack/ods-jenkins-shared-library#473]
**Describe the bug**
We have quite a few users that have reported the following unexpected behaviour when deploying their component to QA using Release Manager in ODS 3.x.

Given a component that requires a new parameter in the `template.yaml`, for instance `SAML_SETTINGS_PATH`:

```
apiVersion: template.openshift.io/v1
kind: Template
objects:
- apiVersion: apps.openshift.io/v1
  kind: DeploymentConfig
  ...
  spec:
    replicas: 1
    ...
      spec:
        containers:
        - env:
          - name: SAML_SETTINGS_PATH
            value: ${SAML_SETTINGS_PATH}
...

parameters:
...

- name: SAML_SETTINGS_PATH
  required: true
```

**To Reproduce**
1. Create an app component through the Provisoning App.
2. Create the Release Manager component through the Provisioning App.
3. Configure Release Manager to deploy the app component in the metadata.yml.
4. Create 3 files in the repository of the Release Manager, `dev.env`, `qa.env` and `prod.env` in the `release/ods-generated-${version}` branch and set any value to `SAML_SETTINGS_PATH`. 
5. In Jenkins, go to the MRO pipeline and deploy a new release version first to DEV, then to QA.

**Expected behavior**
The deployment of the new version to QA using Release Manager should be able to find the parameter `SAML_SETTINGS_PATH` in the corresponding `qa.env` file defined in the Release Manager repository. However, the error described further below occurs.

**Affected version (please complete the following information):**
 - OpenShift: 3.11
 - OpenDevStack 3.x

**Log Output (ensure to remove any confidential information like tokens, project names, etc.**
```
+ tailor --non-interactive -n ****-test apply --selector app=***** --param ODS_OPENSHIFT_APP_DOMAIN=****** --preserve bc:/spec/output/imageLabels --preserve bc:/spec/output/to/name --ignore-unknown-parameters bc,is
error: unable to process template
  Required value: template.parameters[2]: parameter SAML_SETTINGS_PATH is required and must be specified
Comparing templates in . with OCP namespace ***-test.
Limiting resources to bc,is with selector app=****.
Could not process template.yml template: exit status 1
```

**Additional context**

Release Manager is not able to find the param `SAML_SETTINGS_PATH` from the file `qa.env` in the Release Manager repository:

### Component: release-manager
```
Branch: release/ods-generated-1
Files: ./dev.env
	 ./qa.env
	 ./prod.env
```

However, if the param `SAML_SETTINGS_PATH` is **also** specified in the app component in the file `./openshift/{project-key}-qa.env`, Release Manager is able to find the param and the deployment is carried out successfully.

### Component: my-app
```
Branch: release/ods-generated-1
Files: ./openshift/{project-key}-dev.env
	 ./openshift/{project-key}-qa.env
	 ./openshift/{project-key}-prod.env
```

In conclusion, the `SAML_SETTINGS_PATH`  param needs to exist in **both** locations and it is annoning for the users to keep the values of these files in sync.

⚠ *P.S: Note that, to make matters worse, in the app component the naming convention of the `.env` files is different compared to the naming used in the Release Manager component.*

**Comment by michaelsauter** [comment]
Hmm https://github.com/opendevstack/ods-jenkins-shared-library/blob/master/src/org/ods/orchestration/util/Project.groovy#L460 should find the file, and https://github.com/opendevstack/ods-jenkins-shared-library/blob/master/src/org/ods/orchestration/phases/DeployOdsComponent.groovy#L108 should pass it to the invocation, and finally https://github.com/opendevstack/ods-jenkins-shared-library/blob/master/src/org/ods/services/OpenShiftService.groovy#L138 should add it to the `tailor` command. 

Do you have any insight into where in this chain it does not work?

One note regarding the different naming: Yes this is unfortunate. It comes from the fact that Tailor itself has no idea what "an environment" is- Tailor only knows namespaces. That is why Tailor searches for a `<namespace>.env` file in the working dir, where `namespace` is the namespace against which Tailor is executed. Tailor cannot search for `dev.env` or the like because it does not know which env to look for. What we could do is to change the file names in the release manager to be named the same to avoid confusion?

**Comment by felipecruz91** [comment]
@michaelsauter I'll try to investigate it but it will take some time as I am not aware of a quick approach to debug it.

Thanks for taking the time clarifying the different file naming. I couldn't agree more that we should keep the same file naming convention to avoid confusion.

**Comment by metmajer** [comment]
@michaelsauter kudos for always taking the time to help and unblock, we all appreciate a lot!

**Comment by michaelsauter** [comment]
@felipecruz91 My gut feeling is that it doesn't find the file. Maybe something with that workspace variable is wrong here? E.g. some master/agent issue. You could try to add a debug statement if the file is not found maybe?

**Comment by felipecruz91** [comment]
@michaelsauter It has been confirmed that, for ODS 3.x, the `.env` files are only needed in the `openshift` folder (and not in the Release Manager component as it used to be in ODS 2.x).

### Component: my-app
```
Branch: release/ods-generated-1
Files: ./openshift/{project-key}-dev.env
       ./openshift/{project-key}-test.env    <--- notice is `test`, not `qa`
       ./openshift/{project-key}-prod.env
```
It is important to highlight that for this to work when deploying to the QA environment, the file must be named `{project-key}-test.env` instead of `{project-key}-qa.env`.

**Comment by felipecruz91** [comment]
@michaelsauter The [documentation](https://www.opendevstack.org/ods-documentation/opendevstack/3.x/jenkins-shared-library/orchestration-pipeline.html#_customizing_release_bundle_creation) mixes up both namings.

![image](https://user-images.githubusercontent.com/15997951/96870010-92664280-1470-11eb-811b-07eaaba52f48.png)

vs

![image](https://user-images.githubusercontent.com/15997951/96870075-a6aa3f80-1470-11eb-8830-8142ec5106db.png)

**Comment by michaelsauter** [comment]
@felipecruz91 The example you give looks correct to me? If you have `<namespace>.env`, then it is `foo-test` because that's the name of the namespace. If you have the file in the release manager, it should be `qa.env` because `qa` is the name of the environment. I still don't understand why having e.g. `qa.env` in the release manager doesn't work. It should.

**Comment by felipecruz91** [comment]
@michaelsauter What I intend to report here is that, if you try to deploy a component to QA, the deployment fails in Jenkins if the file is  `./openshift/{project-key}-qa.env`.

However, if the file specified is ` ./openshift/{project-key}-test.env`, the deployment to QA succeeds using Release Manager in 3.x.

(I understand that this is not the expected behaviour so that's the reason why I'm reporting it.)

**Comment by michaelsauter** [comment]
If the component repo contains `./openshift/{project-key}-qa.env`, and the contents of that file are required to process the template, then it will fail. That is expected behaviour, as the target project is named `{project-key}-test`, not `{project-key}-qa`. The file needs to be `./openshift/{project-key}-test.env`, as it is also documented in the screenshot you posted above. If this is the only "problem", then I think we can close this issue.

Separate from this is that if a file named `qa.env` exists in the release manager, it should be picked up when deploying to QA. I understand your original issue such that this file is NOT picked up even though it exists. That would be unexpected and a bug.

**Comment by felipecruz91** [comment]
> Separate from this is that if a file named qa.env exists in the release manager, it should be picked up when deploying to QA. I understand your original issue such that this file is NOT picked up even though it exists. That would be unexpected and a bug.

Unfortunately, that's the case. The param does not get picked up if it is *only* specified in the Release Manager repository, as shown below:

```
error: unable to process template
  Required value: template.parameters[2]: parameter PRUEBA is required and must be specified
```

![image](https://user-images.githubusercontent.com/15997951/96882981-28559980-1480-11eb-8176-3160bd0707c9.png)

![image](https://user-images.githubusercontent.com/15997951/96883303-90a47b00-1480-11eb-9ccc-1471ef897fc3.png)

![image](https://user-images.githubusercontent.com/15997951/96883472-bf225600-1480-11eb-95e4-ca45d9df7b25.png)

**Comment by metmajer** [comment]
@felipecruz91 @michaelsauter do you have a playground to look at it together? Consider that 3.1 is coming up in the next weeks. Cheers!

**Comment by felipecruz91** [comment]
@metmajer Yes, we have a safe project where we are trying this out.

@michaelsauter Ping me on Teams if you need to have a look at this together.

**Comment by michaelsauter** [comment]
OK, @felipecruz91 and I went through this, and the issue is that the implementation is quite confusing (but "working").

There are basically two cases to consider:

1. Users do not care about OpenShift templates, so they do not have an `openshift` folder in their repo. Therefore, the deployment from within the component `Jenkinsfile` does not care about live OpenShift resources. In the FINALIZE phase of the orchestration pipeline, the live OpenShift resources are exported. After the export has been done, the `dev.env` file of the release manager is applied **in reverse**, meaning all values in the export matching the **variable values** in the `dev.env` file are replaced with the **variable names**. When the orchestration pipeline runs against QA or PROD, then the `qa.env` and `prov.env` files are used to lookup the values for those envs.

2. Users do care about OpenShift templates, so they have an `openshift` folder in their repo. Therefore, the deployment from within the component `Jenkinsfile` uses the templates in the repo (and the `.env` files in the repo!) to control the live resources. Note that the `dev.env` file from the release manager is NOT used at all this this case ... the component does not know about the release manager, nor does it need this reverse-applying hack. However, when the orchestration pipeline runs against QA or PROD, then the `qa.env` and `prov.env` files are (also) used to lookup the values for those envs.

The confusion is pretty much what #479 describes. The orchestration pipeline uses different means to deploy, depending on which target project is used: it uses the component's `Jenkinsfile` for DEV, and its own logic for QA/PROD.

So, with this in mind:

1. For components where you control OpenShift resources, store the `.env` files in the component repo
2. For components where you export OpenShift resources, use the `.env` files in the release manager

In general, exporting OpenShift resources has serious limitations and potential for all kinds of bugs in the process. Personally, I strongly discourage using it.

I'm closing this as it's not a bug, but we should try to adjust the design so that it is less complex and easier to understand for our users.
